File: landrumower/firmware/lib/buzzer.py
from machine import Pin, PWM
import time

LOUDNESS = 65000

class Buzzer:
    mainUnitState: str = ""
    sound: bool = False
    playPattern: list[int] = []
    currentTime: int = 0
    loudness: int = 0
    nextRunTime: int = 0
    runFrequency: int = 20 #50Hz

    def __init__(self, freq: int, pinPwm: PWM) -> None:
        self.freq = freq
        self.pinPwm = pinPwm

        pinPwm.freq(freq)
    
    def stopPlaying(self) -> None:
        self.sound = False
        self.loudness = 0
        self.playPattern = []
        self.currentTime = 0

    def playInfo(self, loudness: int) -> None:
        self.stopPlaying()
        self.loudness = loudness
        self.playPattern = [1000]
    
    def playImuCalibration(self, loudness: int) -> None:
        self.stopPlaying()
        self.loudness = loudness
        self.playPattern = [100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000, 100]

    def playWanrning(self, loudness: int) -> None:
        self.stopPlaying()
        self.loudness = loudness
        self.playPattern = [500, -1000, 500, -1000, 500]
    
    def playShutdown(self, loudness: int) -> None:
        self.stopPlaying()
        self.loudness = loudness
        self.playPattern = [100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000] 
    
    def checkPlayPattern(self, mainUnitState: str) -> None:
        self.mainUnitState = mainUnitState
        if self.mainUnitState == "boot" or self.mainUnitState == "idle" or self.mainUnitState == "charge" or self.mainUnitState == "dock" or self.mainUnitState == "gps recovery" or self.mainUnitState == "gps wait fix" or self.mainUnitState == "gps wait float":
            self.playInfo(LOUDNESS)
        elif self.mainUnitState == "imu calibration":
            self.playImuCalibration(LOUDNESS)
        elif self.mainUnitState == "mow" or self.mainUnitState == "error" or self.mainUnitState == "escape forward" or self.mainUnitState == "escape reverse":
            self.playWanrning(LOUDNESS)
    
    def run(self, mainUnitState: str) -> None:
        if (time.ticks_diff(time.ticks_ms(), self.nextRunTime)) > 0:
            self.nextRunTime = time.ticks_add(time.ticks_ms(), self.runFrequency)
            if self.mainUnitState != mainUnitState and self.mainUnitState != "shutdown":
                self.checkPlayPattern(mainUnitState)
            if time.ticks_diff(self.currentTime, time.ticks_ms()) < 0:
                if self.playPattern == []:
                    self.stopPlaying()
                elif self.playPattern[0] > 0:
                    self.sound = True
                    self.currentTime = time.ticks_add(time.ticks_ms(), abs(self.playPattern[0]))
                    self.playPattern.pop(0)
                else:
                    self.sound = False
                    self.currentTime = time.ticks_add(time.ticks_ms(), abs(self.playPattern[0]))
                    self.playPattern.pop(0)

            if self.sound:
                self.pinPwm.duty_u16(self.loudness)
            else:
                self.pinPwm.duty_u16(0)

File: landrumower/firmware/3.x/lib/buzzer/buzzer.cpp
#include <Arduino.h>
#include "config.h"
#include "buzzer.h"

Buzzer::Buzzer(uint8_t pin)
{
    this->pin = pin;
}

void Buzzer::setup()
{
    pinMode(pin, OUTPUT);
}

void Buzzer::stopPlaying()
{
    sound = false;
    loudness = 0;
    playPattern.clear();
    currentTime = 0;
}

void Buzzer::playInfo(int loudness)
{
    stopPlaying();
    this->loudness = loudness;
    playPattern = {1000};
}

void Buzzer::playImuCalibration(int loudness)
{
    stopPlaying();
    this->loudness = loudness;
    playPattern = {
        100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000,
        100, -1000, 100, -1000, 100, -1000, 100, -1000, 100, -1000,
        100, -1000, 100};
}

void Buzzer::playWarning(int loudness)
{
    stopPlaying();
    this->loudness = loudness;
    playPattern = {500, -1000, 500, -1000, 500};
}

void Buzzer::playShutdown(int loudness)
{
    stopPlaying();
    this->loudness = loudness;
    playPattern = {
        100, -20, 100, -3000, 100, -20, 100, -3000, 100, -20,
        100, -3000, 100, -20, 100, -3000, 100, -20, 100, -3000,
        100, -20, 100, -3000, 100, -20, 100, -3000};
}

void Buzzer::checkPlayPattern(const String &state)
{
    mainUnitState = state;

    if (state == "boot" || state == "idle" || state == "charge" || state == "dock" ||
        state == "gps recovery" || state == "gps wait fix" || state == "gps wait float")
    {
        playInfo(BUZZER_LOUDNESS);
    }
    else if (state == "imu calibration")
    {
        playImuCalibration(BUZZER_LOUDNESS);
    }
    else if (state == "mow" || state == "error" || state == "escape forward" || state == "escape reverse")
    {
        playWarning(BUZZER_LOUDNESS);
    }
}

void Buzzer::run(const String &state)
{
    int now = millis();

    if ((now - nextRunTime) >= 0)
    {
        nextRunTime = now + runFrequency;

        if (mainUnitState != state && mainUnitState != "shutdown")
        {
            checkPlayPattern(state);
        }

        if ((now - currentTime) >= 0)
        {
            if (playPattern.empty())
            {
                stopPlaying();
            }
            else
            {
                int duration = playPattern.front();
                playPattern.erase(playPattern.begin());
                sound = duration > 0;
                currentTime = now + abs(duration);
            }
        }

        if (sound)
        {
            //digitalWrite(pin, HIGH);
            analogWrite(pin, loudness);
        }
        else
        {
            //digitalWrite(pin, LOW);
            analogWrite(pin, 0);
        }
    }
}